Add control parameters to analytics dashboard

Damonamajor opened the pull request:
This adds 5 params to the homeval report to help produce more refined queries.
They are implemented both in the initial download and caching stage as well as the report.
- start_date: initial date of data or a defined date.
-  excluded_dates: List of dates which can be removed from the query.
-  end_date: date that default to today unless otherwise specified.
- refresh_today: TRUE FALSE indicator whether to query todays data.
- ip_addresses_to_filter: List of IPs which can be removed from the query.

There is also a lintr styler circular error again which I just added #nolint to.

Damonamajor commented on reports/analytics/analytics_dashboard.qmd:
````diff
@@ -52,10 +81,7 @@ if (!file.exists(cache_file)) {
-tz_local <- "America/Chicago"
-baseline_date <- as.Date("2025-09-04")
-
+```{r download_logs}
````
Just added headers to the different sections.

Damonamajor commented on reports/analytics/analytics_dashboard.qmd:
```diff
@@ -173,17 +214,51 @@ walk(dates_to_fetch, function(d) {
+  dplyr::filter(
+    # End date filter
+    if (!is.null(params$end_date) &
+      nzchar(params$end_date)) { # nolint
```
This was giving a circular lint and styler error.

jeancochrane commented on reports/analytics/analytics_dashboard.qmd:
```diff
@@ -173,17 +214,51 @@ walk(dates_to_fetch, function(d) {
+      !(date %in% as.Date(params$excluded_dates))
+    } else {
+      TRUE
+    }
```
**[Suggestion, optional]** It strikes me that we're adding a lot of extra logic to this `filter()` call to account for the possibility that these parameters are missing/empty, when the `convert_params` block has already defined fallback values for the missing/empty case. Why not update the `convert_params` block to process `excluded_dates`, have `emtpy_dates` fall back to an empty vector, and then simplify this filter so that it incorporates the fallback values? Then the filter would look something like:

```r
filter(
    date >= start_date,
    date <= end_date,
    !(date %in% excluded_dates)
)
```

jeancochrane commented on reports/analytics/analytics_dashboard.qmd:
```diff
@@ -4,6 +4,17 @@ execute:
+  start_date:
+  excluded_dates:
+  end_date:
+  refresh_today:
```
**[Suggestion, non-blocking]** The fallback value for this parameter is relatively simple, so I would suggest we add it here, which will also help signpost the format of this param for the user:


```suggestion
  refresh_today: true
```

jeancochrane commented on reports/analytics/analytics_dashboard.qmd:
```diff
@@ -142,13 +165,31 @@ cached_dates <- str_match(
+  setdiff(
+    all_dates[
+      (all_dates != end_date) |
+        (is.null(params$refresh_today) || isTRUE(params$refresh_today))
```
TK: Do we need this if we're going to add it below?

jeancochrane commented on reports/analytics/analytics_dashboard.qmd:
```diff
@@ -159,7 +200,7 @@ walk(dates_to_fetch, function(d) {
   message("Target file: ", fs::path_abs(out_file))
 
-  if (fs::file_exists(out_file) && d != today_local) {
+  if (fs::file_exists(out_file) & d != end_date) {
```
TK: Is this right?

jeancochrane commented on reports/analytics/analytics_dashboard.qmd:
```diff
@@ -173,17 +214,51 @@ walk(dates_to_fetch, function(d) {
+logs_to_import <- tibble(file = logs, date = log_dates) %>%
+  dplyr::filter(
+    # End date filter
+    if (!is.null(params$end_date) &
```
**[Nitpick, required]** The single-ampersands in this filter expression are raising syntax errors for me. I think we want to use double-ampersand because these are not vectorized comparisons, right?


```suggestion
    if (!is.null(params$end_date) &&
```

jeancochrane reviewed (changes requested):
This is trending in the right direction, but some of the code is a bit confusing and I'm not sure if it's working as intended. Let's get the intended behavior sorted out and then we can iron out the details.

Damonamajor commented:
Thanks for the feedback. There is still the nolint included in the file, but I think cleaning the params up the script improves and simplified it a lot. I also updated the refresh today to only work if the date is today. I think that should fix it from working on the end date of other queries.

`  if (d == today(tzone = tz_local) & refresh_today == FALSE &`

Damonamajor commented on reports/analytics/analytics_dashboard.qmd:
```diff
@@ -142,13 +165,31 @@ cached_dates <- str_match(
+    )
+  ),
+  if (is.null(params$refresh_today) ||
+    isTRUE(params$refresh_today)) { # nolint
```
I have it so that it only uses the feature if the date is today's date, so that should help fix this behavior.

jeancochrane reviewed (approved):
Awesome job here! I think the final code is nice and simple, and the filters worked great in every combination I tested.
